src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -133,111 +133,75 @@ private unsafe ValueTask<int> ReadAsyncInternal(Memory<byte> destination, Cancel
+                    nativeOverlapped->OffsetLow = unchecked((int)positionBefore);
+                    nativeOverlapped->OffsetHigh = (int)(positionBefore >> 32);
 
-            // queue an async ReadFile operation and pass in a packed overlapped
-            int r = FileStreamHelpers.ReadFileNative(_fileHandle, destination.Span, false, intOverlapped, out int errorCode);
-
-            // ReadFile, the OS version, will return 0 on failure.  But
-            // my ReadFileNative wrapper returns -1.  My wrapper will return
-            // the following:
-            // On error, r==-1.
-            // On async requests that are still pending, r==-1 w/ errorCode==ERROR_IO_PENDING
-            // on async requests that completed sequentially, r==0
-            // You will NEVER RELIABLY be able to get the number of bytes
-            // read back from this call when using overlapped structures!  You must
-            // not pass in a non-null lpNumBytesRead to ReadFile when using
-            // overlapped structures!  This is by design NT behavior.
-            if (r == -1)
-            {
-                // For pipes, when they hit EOF, they will come here.
-                if (errorCode == Interop.Errors.ERROR_BROKEN_PIPE)
-                {
-                    // Not an error, but EOF.  AsyncFSCallback will NOT be
-                    // called.  Call the user callback here.
-
-                    // We clear the overlapped status bit for this special case.
-                    // Failure to do so looks like we are freeing a pending overlapped later.
-                    intOverlapped->InternalLow = IntPtr.Zero;
-                    valueTaskSource.ReleaseNativeResource();
-                    TryToReuse(valueTaskSource);
-                    return new ValueTask<int>(0);
+                    // When using overlapped IO, the OS is not supposed to
+                    // touch the file pointer location at all.  We will adjust it
+                    // ourselves, but only in memory. This isn't threadsafe.
+                    _filePosition += destination.Length;
                 }
-                else if (errorCode != Interop.Errors.ERROR_IO_PENDING)
-                {
-                    if (!_fileHandle.IsClosed && CanSeek)  // Update Position - It could be anywhere.
-                    {
-                        _filePosition = positionBefore;
-                    }
 
-                    valueTaskSource.ReleaseNativeResource();
-                    TryToReuse(valueTaskSource);
-
-                    if (errorCode == Interop.Errors.ERROR_HANDLE_EOF)
-                    {
-                        ThrowHelper.ThrowEndOfFileException();
-                    }
-                    else
+                // Queue an async ReadFile operation.
+                if (Interop.Kernel32.ReadFile(_fileHandle, (byte*)vts._memoryHandle.Pointer, destination.Length, IntPtr.Zero, nativeOverlapped) == 0)
```

Comment:
> this method should always return 0:

I don't believe that's true.  It can complete synchronously and return 1, but it still posts to the completion port to signal the completion, which is why FILE_SKIP_COMPLETION_PORT_ON_SUCCESS exists.